xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -426,6 +496,111 @@ private String processRouteConfig(RouteConfiguration config) {
     return null;
   }
 
+  /**
+   * Handles EDS response, which contains a list of ClusterLoadAssignment messages. Each
+   * ClusterLoadAssignment message contains information for endpoints in a cluster, grouped
+   * by localities. The response is NACKed if contains invalid information for gRPC's usage.
+   * Otherwise, an ACK request is sent to management server. Information in ClusterLoadAssignment
+   * messages are saved in local cache, in case of corresponding endpoint watchers are registered
+   * later. Endpoint watchers interested in endpoint information contained in this response are
+   * notified.
+   */
+  private void handleEdsResponse(DiscoveryResponse edsResponse) {
+    logger.log(Level.FINE, "Received an EDS response: {0}", edsResponse);
+    checkState(adsStream.edsResourceNames != null,
+        "Never requested for EDS resources, management server is doing something wrong");
+    adsStream.edsRespNonce = edsResponse.getNonce();
+
+    // Unpack ClusterLoadAssignment messages.
+    List<ClusterLoadAssignment> clusterLoadAssignments =
+        new ArrayList<>(edsResponse.getResourcesCount());
+    try {
+      for (com.google.protobuf.Any res : edsResponse.getResourcesList()) {
+        clusterLoadAssignments.add(res.unpack(ClusterLoadAssignment.class));
+      }
+    } catch (InvalidProtocolBufferException e) {
+      adsStream.sendNackRequest(ADS_TYPE_URL_EDS, adsStream.edsResourceNames,
+          edsResponse.getNonce(), "Broken EDS response");
+      return;
+    }
+
+    String errorMessage = null;
+    // Endpoint information updates for requested clusters received in this EDS response.
+    Map<String, EndpointUpdate> endpointUpdates = new HashMap<>();
+    // Walk through each ClusterLoadAssignment message. If any of them for requested clusters
+    // contain invalid information for gRPC's load balancing usage, the whole response is rejected.
+    validateData:
```

Comment:
If you had created a helper function for the following `for` loop you wouldn't need the label and labeled break. Also the code is easier to follow/review/maintain etc

Reply:
Same as we discussed before. Putting things in the same method is more straightforward. Using helper method doesn't help get rid of the label. The label can be easily eliminated by adding another `if` to `break` again after the inner loop. I can change to that style, not much difference.